stdlib/option.gr

```diff
@@ -0,0 +1,129 @@
+# Standard library for list functionality
+
+export *
+
+let isSome = (opt) => {
+  match (opt) {
+    | Some(_) => true
+    | None => false
+  }
+}
+
+let isNone = (opt) => {
+  match (opt) {
+    | None => true
+    | Some(_) => false
+  }
+}
+
+let contains = (val, opt) => {
+  match (opt) {
+    | Some(x) => x == val
+    | None => false
+  }
+}
+
+let expect = (msg, opt) => {
+  match (opt) {
+    | Some(x) => x
+    | None => fail msg
+  }
+}
+
+let unwrap = (opt) => {
+  expect('Could not unwrap None value', opt)
+}
+
+let unwrapWithDefault = (default, opt) => {
+  match (opt) {
+    | Some(x) => x
+    | None => default
+  }
+}
+
+let map = (fn, opt) => {
+  match (opt) {
+    | Some(x) => Some(fn(x))
+    | None => None
+  }
+}
+
+let mapWithDefault = (default, fn, opt) => {
+  match (opt) {
+    | Some(x) => fn(x)
+    | None => default
+  }
+}
+
+# TODO: Do you like `andThen` or should it be `flatMap` or both?
+let andThen = (fn, opt) => {
+  match (opt) {
+    | Some(x) => fn(x)
+    | None => None
+  }
+}
+
+let filter = (pred, opt) => {
+  match (opt) {
+    | Some(x) =>
+      if (pred(x)) {
+        Some(x)
+      } else {
+        None
+      }
+    | None => None
+  }
+}
+
+let zip = (optA, optB) => {
+  match ((optA, optB)) {
+    | (Some(a), Some(b)) => Some((a, b))
+    | (Some(_), None) => None
+    | (None, Some(_)) => None
+    | (None, None) => None
+  }
+}
+
+let zipWith = (fn, optA, optB) => {
+  match ((optA, optB)) {
+    | (Some(a), Some(b)) => Some(fn(a, b))
+    | (Some(_), None) => None
+    | (None, Some(_)) => None
+    | (None, None) => None
+  }
+}
+
+# TODO: Would be cool to have a recursive version
```

Comment:
I don't think the typechecker would let us write one without a custom type to match against :/

The type would have to look like this:
```grain
data RecOpt<a> = RecOptCons(Option<a>, RecOpt<a>) | RecOptNone
```

You could maybe get this to work with higher kinded types, but I'm not sure.

Reply:
Yeah, I'm not sure how calls to a recursive version would type-check in the first place...